### PR TITLE
Gradient circuits 5/n: refactor base Differentiator

### LINKED ISSUE #516: Refactor differentiation logic to use get_gradient_circuits

### PR DESCRIPTION
**Pull request by zaqqwerty**
Resolves #516 .  Now the child classes of `Differentiator` only need to implement `get_gradient_circuits`, the base class uniformly handles turning those outputs into derivatives.

### CODE REVIEW

**Comment by MichaelBroughton** [comment]
After offline discussions: Also need to update docstring and fix interfaces so that if the user defines `get_gradient_circuits` we guess at an implementation of `differentiate_analytic` and `differentiate_sampled` if they didn't already define both for us. Also need to make it clear in docstring about new interface changes.

**Review by jaeyoo — APPROVED** [review]
LGTM

**Review by jaeyoo — CHANGES_REQUESTED** [review]
Please fix the error

File: tensorflow_quantum/python/differentiators/linear_combination.py [hunk 1]
```diff
@@ -111,7 +111,6 @@ def __init__(self, weights, perturbations):
         self.n_non_zero_perturbations = tf.gather(
             tf.shape(self.non_zero_perturbations), 0)
 
-    @tf.function
```
Comment (MichaelBroughton) [inline tensorflow_quantum/python/differentiators/linear_combination.py:114]: Think we might want to keep this ?

Comment (zaqqwerty) [reply tensorflow_quantum/python/differentiators/linear_combination.py]: Added it back

File: docs/tutorials/gradients.ipynb [hunk 2]
```diff
@@ -604,9 +604,20 @@
       "source": [
         "## 4. Advanced usage\n",
         "Here you will learn how to define your own custom differentiation routines for quantum circuits.\n",
-        "All differentiators that exist inside of TensorFlow Quantum subclass `tfq.differentiators.Differentiator`. A differentiator must implement `differentiate_analytic` and `differentiate_sampled`.\n",
+        "All differentiators that exist inside of TensorFlow Quantum subclass `tfq.differentiators.Differentiator`. A differentiator must implement `get_gradient_circuits`, which tells the base class which circuits to measure to obtain an estimate of the gradient.\n",
```
Comment (MichaelBroughton) [inline docs/tutorials/gradients.ipynb:607]: Might want to note here that you can also (optionally) overload `differentiate_analytic` or `differentiate_sampled`.

**Review by MichaelBroughton — APPROVED** [review]
LGTM, might want to just clean up the messaging in the tutorial a bit.

File: docs/tutorials/gradients.ipynb (hunk 2 above)
Comment (zaqqwerty) [reply docs/tutorials/gradients.ipynb]: Done

**Review by MichaelBroughton — APPROVED** [review]
LGTM

**Pull request merged**